src/js/service/VendingMachine.js

```diff
@@ -1,67 +1,97 @@
-import { ALERT_MESSAGE } from './constant.js';
-import { removeSpace } from '../util/string.js';
-import { isAmountValid, isNameValid, isPriceValid } from './validator.js';
-import ValidationError from './ValidationError.js';
 import UnitCountMachine from './UnitCountMachine.js';
+import ProductManager from './ProductManager.js';
+import { loadItem } from '../util/dataSaver.js';
+import { DATA_STORAGE } from '../ui/constant.js';
+import { isGreaterThan, isInteger, isMultipleOf } from './validator.js';
+import ValidationError from './ValidationError.js';
+import { ERROR_MESSAGE, VENDING_MACHINE_CONSTANT } from './constant.js';
+
+/**
+ * @typedef {import('../service/ProductManager').ProductItem} ProductItem
+ * @typedef {import('./UnitCountMachine.js').UnitCountInfo} UnitCountInfo
+ */
+
+//prettier-ignore
+const {
+  MIN_AMOUNT,
+  MULTIPLE,
+} = VENDING_MACHINE_CONSTANT.PURCHASABLE_MONEY;
 
 export class VendingMachine {
-  #products;
+  /** @type {ProductManager} */
+  #productManager;
   /**@type {UnitCountMachine} */
   #unitCountMachine;
+  /**@type {number} */
+  #insertedMoney;
 
-  constructor() {
-    this.reset();
-  }
-
-  reset() {
-    this.#products = [];
-    this.#unitCountMachine = new UnitCountMachine();
+  /**
+   * @param {?ProductItem[]} products
+   * @param {?UnitCountInfo} unitCountInfo
+   */
+  constructor(products, unitCountInfo) {
+    this.#productManager = new ProductManager(products);
+    this.#unitCountMachine = new UnitCountMachine(unitCountInfo);
+    this.#insertedMoney = 0;
   }
 
   /**
    *
-   * @returns {VendingMachineItem[]}
+   * @returns {ProductManager}
    */
-  get products() {
-    return this.#products;
+  get productManager() {
+    return this.#productManager;
   }
 
+  /**
+   * @returns {UnitCountMachine}
+   */
   get unitCountMachine() {
     return this.#unitCountMachine;
   }
 
-  /**
-   *
-   * @param {VendingMachineItem} vendingMachineItem
-   */
-  addItem({ name, price, amount }) {
-    this.#validateProduct({ name, price, amount });
-    this.#products = [
-      ...this.#products.filter((item) => item.name !== name),
-      {
-        index: this.#products,
-        name: removeSpace(name),
-        price,
-        amount,
-      },
-    ].sort((a, b) => a.index - b.index);
+  get insertedMoney() {
+    return this.#insertedMoney;
   }
 
   /**
-   *
-   * @param {VendingMachineItem} vendingMachineItem
+   * @param {number} index
+   * @returns {ProductItem}
    */
-  #validateProduct({ name, price, amount }) {
-    if (!isNameValid(name)) {
-      throw new ValidationError(ALERT_MESSAGE.VALIDATION.PRODUCT.NAME_BLANK);
+  purchase(index) {
+    const product = this.#productManager.products[index];
+    if (!(product && this.#insertedMoney - product.price >= 0)) {
```

Comment:
```suggestion
    if (!product || this.#insertedMoney < product.price) {
```
저는 개인적으로 이렇게 표현하는게 더 명확하다고 생각해요!

지금은
- 조건문을 이해해야 하고 ( 제품이 있어야 하고, 자판기에 들어온 돈이 제품의 가격보다 커야 함)
  - 가격 비교도 단순 비교가 아니라 연산 후에 0보다 큰지 봐야하네요.. ㅎㅎ
- 해당 조건의 반대되는 경우 또한 이해해야합니다.

그래서 제안해드린 코드로 변경할 경우
- 제품이 없거나
- 제품 가격보다 작거나

이럴 경우 코드를 이해하는 비용이 더 적어지지 않나요!?

Reply:
코드를 이해하는 비용에 관해서도 여러 모로 고민이 필요한 부분이군요!